### ISSUE TITLE
[SemaStmt]:enumeration of different enum type,but setAllEnumCasesCovered() as True, is this behavior right? 

### ISSUE THREAD

**Issue by mzyKi** [llvm/llvm-project#80115]
```cpp
enum E { e1a, e1b };
enum Other { e2a };

void f(E e){
switch (e) {
  case e1b: // e1b is an enumerator of E
    return;
  case e2a: // e2a is not an enumerator of E
    return;
  }
}
```

In the above case, I found that the result of  ```isAllEnumCasesCovered()``` of this ```switchStmt``` is True. I fell confused about it, is this right as expected?

**Comment by mzyKi** [comment]
@MaskRay  Could you please give me some suggestion?

**Comment by shafik** [comment]
In `Sema::ActOnFinishSwitchStmt(...)` it looks like we collect each enumerator and its value here:

https://github.com/llvm/llvm-project/blob/a03a6e99647318a86ea398c42e241da43e3c550e/clang/lib/Sema/SemaStmt.cpp#L1519-L1525

but then when we are checking if we are covering the range here:

https://github.com/llvm/llvm-project/blob/a03a6e99647318a86ea398c42e241da43e3c550e/clang/lib/Sema/SemaStmt.cpp#L1568-L1626

we only check the enumerator value and not its type. I believe we should check both.

Interesting to note gcc gets this wrong as well: https://godbolt.org/z/GG46Txnfe

Although clang does catch the mismatched type via `-Wenum-compare-switch`

**Comment by mzyKi** [comment]
Thanks for your explanation,but I am still not sure whether we need to check both value and type in SemaStmt? If need, I can try to fix it. @shafik

**Comment by shafik** [comment]
We need to modify the check here, I think:

https://github.com/llvm/llvm-project/blob/a03a6e99647318a86ea398c42e241da43e3c550e/clang/lib/Sema/SemaStmt.cpp#L1590

To be something like this:

```cpp
        bool TypesEqual = true; 
        // Better name than CurrCase
        Expr *CurrCase = CI->second->getLHS()->IgnoreParenCasts();
        if (DeclRefExpr *DeclRef = dyn_cast<DeclRefExpr>(CurrCase)) {
            // This may not be needed, we may always have an EnumConstantDecl
            if(EnumConstantDecl *ECD  = dyn_cast<EnumConstantDecl>(DeclRef->getDecl()))
              TypesEqual = ECD->getType() == EI->second->getType();
        }
        if (CI != CaseVals.end() && CI->first == EI->first && TypesEqual)
          continue;

        // skip code not relevant to the fix

        if (RI == CaseRanges.end() || EI->first < RI->first || !TypesEqual)
```

CC @AaronBallman @cor3ntin to make sure this looks reasonable

**Comment by AaronBallman** [comment]
That looks reasonably correct to me, yup!

**Comment by shafik** [comment]
@mzyKi let me know if you would like me to assign the issue to you.

I outlined a fix, so feel free to take that and open a PR, you will need to add a release note and add a test.

**Comment by zygoloid** [comment]
It seems to me that the switch coverage warning is already correct. We should be checking that there's a `case` label with the value of each enumerator, irrespective of whether it names the enumerator; otherwise we'll start issuing bogus warnings on codebases that are intentionally performing conversions in the `case`s.

Instead, we should warn that the switch is switching on one enum but the case specifies a value whose type is another enum... which again we already do. So it seems to me that there's nothing to fix here.

**Comment by shafik** [comment]
@mzyKi I think the issues brought up are correct, at least from some quick tests I did.

**Comment by mzyKi** [comment]
Okay, I have assigned this issue to myself and create one pull request after.

**Comment by shafik** [comment]
Let me clarify, I think the issues zygoloid brought up are valid. I think it is a bit confusing but but we don't want to break the case where someone explicitly cases an enumerator value to the correct type e.g.:

```cpp
case static_cast<E>(e2a): 
```

and my proposed fix will flag this case as well, which is not correct.
